testing/example_scripts/issue_519.py

```diff
@@ -22,13 +22,13 @@ def checked_order():
     assert order == [
         ("issue_519.py", "fix1", "arg1v1"),
         ("test_one[arg1v1-arg2v1]", "fix2", "arg2v1"),
-        ("test_two[arg1v1-arg2v1]", "fix2", "arg2v1"),
```

Comment:
As I said in https://github.com/pytest-dev/pytest/pull/11231/files#r1278270549 IMO this change makes sense (though @RonnyPfannschmidt may still disagree).

But I would like to understand what exactly in this PR is causing this change? @sadra-barikbin do you think you can explain it?

Reply:
i left another comment, i think we can choose to opt for the new ordering if we ensure to add the tools i mentioned there eventually

Reply:
Current ordering stems from the way `add_funcarg_pseudo_fixture_def` assigns index to params. It assigns `i` to the index of `all` of the params in `i`th callspec of the metafunc. In our example, `test_one` and `test_two`, each have 4 callspecs and two params (`arg1` and `arg2`). We assign index 0 to `arg1` and `arg2` of `test_one[arg1v1,arg2v1]` and also to those of `test_two[arg1v1,arg2v1]`. We assign index 1 to `arg1` and `arg2` of `test_one[arg1v1,arg2v2]` and also to those of `test_two[arg1v1,arg2v2]` and so on. As a result, `i`th callspecs of the two metafuncs would have common fixture keys.
The new way assigns `i` to the index of param `j` in all callspecs that use `i`th value of `j`. So the decision is made on a per-param basis. In our example, we would assign index 0 to `arg1` of the items that use `arg1v1` and index 1 to `arg1` of the items that use `arg1v2` and so on. As a result, if two callspecs have the `i`th value of the param `j`, they have a common fixture key.

The new way is improved in determining what `i`th value of the param is when there are multiple parametersets in #11257 .

Reply:
After reading your comment and looking more closely at the issue_519.py file, I think I understood some things:

---

First, I'm really not used to the `pytest_generate_tests` hook, so I rewrote the test using `pytest.mark.parametrize`. I also changed some of the names to make things clearer to me:
 
<details>
<summary>Rewritten issue_519.py</summary>

```py
import pprint
from typing import List
from typing import Tuple

import pytest


@pytest.fixture(scope="session")
def checked_order():
    order: List[Tuple[str, str, str]] = []

    yield order
    pprint.pprint(order)


@pytest.fixture(scope="module")
def fixmod(request, a, checked_order):
    checked_order.append((request.node.name, "fixmod", a))
    yield "fixmod-" + a


@pytest.fixture(scope="function")
def fixfun(request, fixmod, b, checked_order):
    checked_order.append((request.node.name, "fixfun", b))
    yield "fixfun-" + b + fixmod


@pytest.mark.parametrize("b", ["b1", "b2"], scope="function")
@pytest.mark.parametrize("a", ["a1", "a2"], scope="module")
def test_one(fixfun, request):
    print()
    # print(request._pyfuncitem.nodeid, request._pyfuncitem.callspec)


@pytest.mark.parametrize("b", ["b1", "b2"], scope="function")
@pytest.mark.parametrize("a", ["a1", "a2"], scope="module")
def test_two(fixfun, request):
    print()
    # print(request._pyfuncitem.nodeid, request._pyfuncitem.callspec)
```
</details>

This made me realize that the apparent niceness of the previous ordering, which orders `test_one[arg1v1-arg2v1]` next to `test_two[arg1v1-arg2v1]`, thus saving a setup, is pretty arbitrary. This can be shown by making the values of the function-scoped parameter (`arg2`/`b`) different in `test_one` and `test_two`:

```diff
-@pytest.mark.parametrize("b", ["b1", "b2"], scope="function")
+@pytest.mark.parametrize("b", ["b11", "b12"], scope="function")
 @pytest.mark.parametrize("a", ["a1", "a2"], scope="module")
 def test_one(fixfun, request):
     print()
     # print(request._pyfuncitem.nodeid, request._pyfuncitem.callspec)
 
 
-@pytest.mark.parametrize("b", ["b1", "b2"], scope="function")
+@pytest.mark.parametrize("b", ["b21", "b22"], scope="function")
 @pytest.mark.parametrize("a", ["a1", "a2"], scope="module")
 def test_two(fixfun, request):
     print()
```

The ordering stays the same, which now doesn't make sense.

---

I also figured that the entire `param_index` thing was much more useful/natural in the pre-`parametrize` times, when only `pytest_generate_tests`/`metafunc` was used for parametrizing. With `pytest_generate_tests` you often use the same parameter set for different tests, as is done in the issue_519.py file. But with `parametrize`, which is mostly done for a single test-function only, it is much less common, I think.